src/invidious/database/migrator.cr

```diff
@@ -46,4 +53,32 @@ class Invidious::Database::Migrator
       )
     SQL
   end
+
+  private def back_up_database
+    table_names_request = <<-SQL
+      SELECT tablename FROM pg_catalog.pg_tables
+      WHERE schemaname = 'public'
+    SQL
+
+    table_names = @db.query_all(table_names_request, as: String)
+
+    table_names.try &.each do |name|
+      copy_table(name)
+    end
+  end
+
+  private def copy_table(table_name : String)
+    @db.exec <<-SQL
+      CREATE TABLE IF NOT EXISTS backup.#{table_name} (
+        id bigserial PRIMARY KEY
+      )
+    SQL
+
+    @db.exec("DROP TABLE backup.#{table_name}")
```

Comment:
The problem I see with that is that it might nuke a perfectly valid backup if invidious is interrupted mid-migration and the user tries to run it again immediately after.

Reply:
so there's stage 1,2,3. the migration can get interrupted pre-1,pre-2,pre-3 & post-3.

pre-1 : The migration wouldn't be run (hopefully upon re-attempting migration the migration is question will still be queued up)

pre-2 : At this point, there exists a table backup.cool_table - it is empty. An interuption now, would leave a empty table that'd be re-populated upon re-attempt

pre-3 : at this point, there is a backup.cool_table and a lack of backup.cool_table (this latter one being cool_tables version right before migration). At this point, critical data is destroyed and we must replenish it ASAP - an interuption now would be fatal

Ok - I see what you're saying, what if I created a copy of backup.cool_table (Pre-Migration's Version) earlier on?

Reply:
The following is based off the assumption that former broquemonsieur would write SQL that looks redundant/hacky but actually is doing the right thing due to SQL's odd and vague limitations where you cannot do something that should be simple or the only way to achieve a simple operation (that seemingly should be a one-liner) is the combo of two ops (a.k.a. two-liner). Present day broquemonsieur has failing memory but can vouch for this ^ :)
---
Meet Sonic, a new Invidious host. At this point in her users have a few playlists saved and Sonic has not migrated her DB...ever. Her users playlists look like this:

| title      | author      |
| ------------- | ------------- |
| Mark Hamill Moments | kilowattwomxn |
| Awesome Trailers | movieDad |

Sonic migrates her DB:
Stage 1 - Creates a table `backup.playlists`. It is empty
Stage 2 - This op done nothing (it fails silently). (If Sonic were to have a backup of her users' playlists already there, it would be destroyed now to make way for a snapshot of her user's present-day playlists) tl;dr This stages deletes the prior version of  `backup.playlists`, not the one you just read out in Stage 1 (Yes, SQL forums agree this is weird that SQL has to be written this way lol)
Stage 3 - This op populates  `backup.playlists` with

| title      | author      |
| ------------- | ------------- |
| Mark Hamill Moments | kilowattwomxn |
| Awesome Trailers | movieDad |
---
Meet Amy, a jaded Invidious host who already has a backup for her users' playlists.

Amy migrates her DB:
you get the idea :)
---

That should clear it up (mainly I was trying to explain for my recollection to myself lol). But [basically](https://github.com/iv-org/invidious/pull/4359#discussion_r1685807683), the system is vulnerable . At time this comment is sent, I begin efforts to prevent memory loss given an asteroid strike at any point in time

Reply:
Hi! Reposting back from my coding session. Now, we only destroy the "second most recent" backup. Whenever an asteroid strikes, the "most recent" backup will always be preserved.

Reply:
Asteroid strikes are useless analogies. 
Meet Dale, he is chilling in Mexican beach propping a MacBook Air on pool tube. He is SSH'd to his Invidious server and tries to do DB migration - a major part of which is DB backup.

The hotel Wifi could go out, the MacBook could get wet or the Server itself could shutdown at any point during this critical time and **still** Dale's users would not lose 

1. "most recent" backup 
2. "2nd most recent" backup
3. regular DB tables